beacon-chain/sync/execution_payload_header.go

```diff
@@ -0,0 +1,145 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/cache"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/verification"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/monitoring/tracing"
+	v1 "github.com/prysmaticlabs/prysm/v5/proto/engine/v1"
+	"go.opencensus.io/trace"
+	"google.golang.org/protobuf/proto"
+)
+
+func (s *Service) validateExecutionPayloadHeader(ctx context.Context, pid peer.ID, msg *pubsub.Message) (pubsub.ValidationResult, error) {
+	if pid == s.cfg.p2p.PeerID() {
+		return pubsub.ValidationAccept, nil
+	}
+
+	if s.cfg.initialSync.Syncing() {
+		return pubsub.ValidationIgnore, nil
+	}
+
+	ctx, span := trace.StartSpan(ctx, "sync.validateExecutionPayloadHeader")
+	defer span.End()
+
+	if msg.Topic == nil {
+		return pubsub.ValidationReject, errInvalidTopic
+	}
+
+	m, err := s.decodePubsubMessage(msg)
+	if err != nil {
+		tracing.AnnotateError(span, err)
+		return pubsub.ValidationReject, err
+	}
+
+	signedHeader, ok := m.(*v1.SignedExecutionPayloadHeader)
+	if !ok {
+		return pubsub.ValidationReject, errWrongMessage
+	}
+	shm := signedHeader.Message
+	slot := shm.Slot
+	builderIndex := shm.BuilderIndex
+
+	if seenBuilderBySlot(slot, builderIndex) {
+		return pubsub.ValidationIgnore, fmt.Errorf("builder %d has already been seen in slot %d", builderIndex, slot)
+	}
+
+	highestValueHeader := cache.SignedExecutionPayloadHeaderByHash(slot, shm.ParentBlockHash)
```

Comment:
What happens in the following scenario, last full block was `N` with the block hash of `H`. Current slot is `N+3`. Our node is about to propose a block for `N+3` and will base it on `N+2`. It requests for the best bid. However, one malicious builder has broadcasted a header with `value: 3000ETH, slot: N+3, parentBlockHash: H` but the parent Block root of `N+2`?